### PR TITLE
ARROW-765: [Python] Add more natural Exception type hierarchy for thirdparty users

### PR DESCRIPTION
**Pull request by wesm**
I also took the liberty of changing a number of error types in libarrow_python

### CODE REVIEW

File: python/pyarrow/error.pyx [hunk 1]
```diff
@@ -19,13 +19,52 @@ from pyarrow.includes.libarrow cimport CStatus
 from pyarrow.includes.common cimport c_string
 from pyarrow.compat import frombytes
 
+
 class ArrowException(Exception):
     pass
 
+
+class ArrowInvalid(ValueError, ArrowException):
```
Comment (jeffknupp) [inline python/pyarrow/error.pyx]: Isn't this an antipattern (deriving from a built-in exception or, worse, raising them directly)? If a user combines some Arrow code with their own in a try block, they may inadvertently catch, say, an `ArrowInvalid` Exception when they explicitly were trying to handle a `ValueError` for non-arrow code in the try block.

tl;dr if you structure your Exceptions like this, client code has no idea from where the issue originated and has no easy way of disambiguating in the `except` block (if they're even aware of the Arrow Exception class hierarchy to begin with).

Comment (wesm) [inline python/pyarrow/error.pyx]: Is it an antipattern? I have never thought there was anything wrong either with

* deriving from built-in exceptions, like ValueError
* raising built-in exceptions directly (pretty much all Python data libraries -- pandas, NumPy, scikit-learn, etc. already do this)

So there's a couple ways to go:

* Don't define user-defined exceptions, raise built-in exceptions only
* Don't derive from a base ArrowException -- which is there to help indicate that the error arose in C++ code
* Derive from base Exception (worst option, IMHO)

adding @jreback for more opinions (I did this at his asking)

Comment (jreback) [inline python/pyarrow/error.pyx]: my main concern was to NOT have the arrow exception hierarchy just be based on ``Exception``. Being able to catch ``ValueError``, ``TypeError``, ``KeyError``, ``IndexError`` (subclass) as errors is very pythonic.

As for the naminghaving ``ArrowTypeError(ArrowException, TypeError)``. I am not how useful that actually is in practice. I would say if you have a need to differentiate between *two* (or more ``TypeError`` s), then having a custom exception is useful (as opposed to having to check the actual error message).

So fine with this PR (though maybe raise the standard python exceptions, unless you have a real need for them).

Comment (jreback) [inline python/pyarrow/error.pyx]: We do have a couple of these in pandas: https://github.com/pandas-dev/pandas/blob/master/pandas/errors/__init__.py

Comment (wesm) [inline python/pyarrow/error.pyx]: My preference is to only raise these exceptions for errors that originate from the C++ libraries, otherwise use the Python built-in ones. Sound good?

Comment (jreback) [inline python/pyarrow/error.pyx]: sgtm.
